sticht/rollbacks/sources/splunk.py

```diff
@@ -116,52 +115,51 @@ def process_result(self, result: Optional[List[Dict[Any, Any]]], earliest_timest
         number of results from a query against configured thresholds to determine
         whether or not to rollback or not
         """
-        if earliest_timestamp_seconds != 'now':
-            if earliest_timestamp_seconds < self.start_timestamp_seconds:
-                self.bad_before_mark = self.is_window_bad(result)
-        else:
+        if earliest_timestamp_seconds > self.start_timestamp_seconds:
             self.bad_after_mark = self.is_window_bad(result)
+        else:
+            self.bad_before_mark = self.is_window_bad(result)
 
         old_failing = self.failing
         self.failing = self.bad_after_mark and not self.bad_before_mark
+
         if self.failing == (not old_failing):
             self.on_failure_callback(self)
 
     def is_window_bad(self, result) -> bool:
+        # Results return an array | Number returns a number
         if self._query_type == 'results':
-            if self._lower_bound and len(result) < self._lower_bound:
-                return True
-            elif self._upper_bound and len(result) > self._upper_bound:
+            result_num = len(result)
+        elif self._query_type == 'number':
+            result_num = next(iter(result[0].values())) if len(result) else len(result)
+
+        if self._lower_bound and self._upper_bound:
```

Comment:
Minor, minor nit/style question: the previous way we did it seemed to deduplicate some of the arithmetic we had to do, checking if each bound existed and then comparing result -- do we need to treat the case where we've specified both a lower and upper bound separately if we have only specified one of them? if in the future there'll be something special about when we have both bounds specified, then makes sense to leave, but if it's only ever gonna be a numerical comparison and we'll be a bad window if we're out of either bound, previous way seemed a bit sleeker.

Reply:
That's a good point. I can only ever think of using the upper and lower bounds separately, but not together. 

We don't specifically test for is_window_bad, but we check if the data was bad before the deployment (bad_before_mark).
https://github.com/Yelp/sticht/blob/master/tests/rollbacks/sources/test_splunk.py#L38

Reply:
The logic is changed from the original because previously we had it mixed up.

`is_window_bad` is checking for data that is specifically bad.

Anything that falls within the bounds, triggers an autorollback. So if data is outside the bounds (i.e. data is lower than lower_bound), then it's considered good data.

For example, a query is searching for any events in the service_stream_errors. Setting the lower_bound to 3 means 0-2 errors are okay and considered normal, but anything beyond that is considered bad. 

I can't think of a use case to keep both lower and upper bounds; however, I'm inclined to keep it to keep with the spirit of giving the user full flexibility to how they want to create their rollback conditions.

Reply:
Ohhh, interesting... so the 'bounds' here aren't for indicating *good* data but for indicating *bad* data?  I was interpreting the bounds as indicating 'good' data, so expected anything *within* the bounds meant the window was 'good', but this is actually the opposite. Good to know!